nucypher/cli/actions/collect.py

```diff
@@ -0,0 +1,198 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from collections import namedtuple
+
+import click
+import maya
+from constant_sorrow.constants import FEDERATED
+from datetime import timedelta
+from typing import Tuple
+from web3.main import Web3
+
+from nucypher.characters.control.emitters import StdoutEmitter
+from nucypher.characters.lawful import Bob, Alice
+from nucypher.cli.painting.help import enforce_probationary_period, paint_probationary_period_disclaimer
+from nucypher.cli.painting.policies import paint_single_card
+from nucypher.cli.types import GWEI
+from nucypher.policy.identity import Card
+
+PublicKeys = namedtuple('PublicKeys', 'encrypting_key verifying_key')
+PolicyParameters = namedtuple('PolicyParameters', 'label m n expiration value rate')
+
+
+def collect_keys_from_card(emitter: StdoutEmitter, card_identifier: str, force: bool):
+    emitter.message(f"Searching contacts for {card_identifier}\n", color='yellow')
+    card = Card.load(identifier=card_identifier)
+
+    if card.character is not Bob:
+        emitter.error('Grantee card is not a Bob.')
+        raise click.Abort
+    paint_single_card(emitter=emitter, card=card)
+
+    if not force:
+        click.confirm('Is this the correct grantee (Bob)?', abort=True)
+
+    bob_encrypting_key = card.encrypting_key.hex()
+    bob_verifying_key = card.verifying_key.hex()
+    public_keys = PublicKeys(encrypting_key=bob_encrypting_key, verifying_key=bob_verifying_key)
+    return public_keys
+
+
+def collect_bob_public_keys(
+        emitter: StdoutEmitter,
+        force: bool,
+        card_identifier: str,
+        bob_encrypting_key: str,
+        bob_verifying_key: str
+        ) -> PublicKeys:
+    """helper function for collecting Bob's public keys interactively in the Alice CLI."""
+
+    # From Bob card
+    if card_identifier and any((bob_encrypting_key, bob_verifying_key)):
+        message = '--bob cannot be used with --bob-encrypting-key or --bob-verifying key'
```

Comment:
Feels like a little leakage here...`--bob` as it relates to `card_identifier`

We could leave it or move verifications to the originating source before we call `collect_bob_public_keys()`. Something to think about...

Reply:
:heavy_check_mark: